judge/ratings.py

```diff
@@ -127,31 +125,41 @@ def rate_contest(contest):
         users = users.exclude(user__rating__lt=contest.rating_floor)
     if contest.rating_ceiling is not None:
         users = users.exclude(user__rating__gt=contest.rating_ceiling)
+
+    def get_user_data(data, user_ids):
+        old_data = [data.get(user, (1200, 535, 0)) for user in user_ids]
+        # rating, volatility, times_ranked
+        return tuple(map(list, zip(*old_data)))
+
     users = list(tie_ranker(users, key=itemgetter(2, 3)))
     participation_ids = [user[1][0] for user in users]
     user_ids = [user[1][1] for user in users]
+    old_rating, old_volatility, times_ranked = get_user_data(data, user_ids)
     ranking = list(map(itemgetter(0), users))
-    old_data = [data.get(user, (1200, 535, 0)) for user in user_ids]
-    old_rating = list(map(itemgetter(0), old_data))
-    old_volatility = list(map(itemgetter(1), old_data))
-    times_ranked = list(map(itemgetter(2), old_data))
     rating, volatility = recalculate_ratings(old_rating, old_volatility, ranking, times_ranked)
 
+    excluded_user_ids = contest.users.filter(virtual=0).exclude(id__in=participation_ids) \
```

Comment:
Note: this is pretty inefficient (since `participation_ids` is almost all participants while `excluded_user_ids` is basically no one). It probably won't matter that much for the current DMOJ participant numbers (~200). We can use `judge.utils.raw_sql.join_sql_subquery` here to make it more efficient, but is it worth it to sacrifice the readability?